src/base/static/components/input-form/input-form.js

```diff
@@ -0,0 +1,833 @@
+        }
+        if (field.type === constants.RICH_TEXTAREA_FIELD_TYPENAME) {
+          this.richTextFields.push(field.name);
+        }
+      });
+
+    this.setState({
+      selectedCategory: categoryConfig.category,
+      isCategoryMenuCollapsed: true,
+      fieldValues: initialFieldStates,
+      isFormSubmitting: false,
+      formValidationErrors: []
+    });
+  }
+
+  resolveFieldType(fieldConfig) {
+    if (fieldConfig.type === constants.COMMON_FORM_ELEMENT_TYPENAME) {
+      return this.props.placeConfig.common_form_elements[fieldConfig.name];
+    } else {
+      return fieldConfig;
+    }
+  }
+
+  getInitialFieldState(fieldConfig) {
+    // "autofill" is a better term than "autocomplete" for this feature.
+    // TODO: update this throughout the codebase.
+    let autofillValue = (fieldConfig.autocomplete)
+      ? Util.getAutocompleteValue(this.resolveFieldType(fieldConfig).name)
+      : null;
+
+    fieldConfig.hasAutofill = !!autofillValue;
+
+    switch(fieldConfig.type) {
+      case constants.BIG_TOGGLE_FIELD_TYPENAME:
+        // NOTE: if binary toggle buttons don't have a saved autofill or
+        // declared default value, assume their default value is the value
+        // associated with the "off" position of the toggle.
+        return autofillValue || fieldConfig.default_value || fieldConfig.content[1].value;
+        break;
+      case constants.BIG_CHECKBOX_FIELD_TYPENAME:
+        return autofillValue || fieldConfig.default_value || [];
+        break;
+      case constants.PUBLISH_CONTROL_TOOLBAR_TYPENAME:
+        return autofillValue || fieldConfig.default_value || "isPublished";
+        break;
+      case constants.COMMON_FORM_ELEMENT_TYPENAME:
+        const commonFormElementConfig = Object.assign(
+          {},
+          this.props.placeConfig.common_form_elements[fieldConfig.name],
+          {name: fieldConfig.name}
+        );
+        return this.getInitialFieldState(commonFormElementConfig);
+        break;
+      default:
+        return autofillValue || fieldConfig.default_value || "";
+        break;
+    }
+  }
+
+  buildField(fieldConfig) {
+
+    // TODO: appropriate field visibility based on admin status
+    // TODO: these field definitions will eventually be factored out for reuse
+    //       in the detail view editor.
+
+    const { autofillMode, emitter, map, mapConfig, placeConfig, router } = this.props;
+    const { fieldValues, isFormSubmitting } = this.state;
+    const classNames = {
+      optionalMsg: cn("input-form__optional-msg", {
+        "input-form__optional-msg--visible": fieldConfig.optional,
+        "input-form__optional-msg--hidden": !fieldConfig.optional
+      })
+    };
+    const fieldPrompt = 
+      <p className="input-form__field-prompt">
+        {fieldConfig.prompt}
+        <span className={classNames.optionalMsg}>
+          {messages.optionalMsg}
+        </span>
+      </p>;
+
+    switch(fieldConfig.type) {
```

Comment:
This switch/case is a code smell that we can benefit from a `field` abstraction.